### PR TITLE
Fix #5456: Keep margins <= .5 in tight_layout

### LINKED ISSUE #5456: 'bottom cannot be >= top' when using `tight_layout`
I'm trying to understand why sometimes I get a `ValueError` when using `tight_layout`.

The code of `matplotlib` is too complicated for me, but I'm trying to see what's happening using `pdb`.

It seems that the problem comes from what `get_tight_layout_figure` returns:

```
(Pdb) l
379                                              subplot_list=subplot_list,
380                                              ax_bbox_list=ax_bbox_list,
381                                              pad=pad, h_pad=h_pad, w_pad=w_pad,
382                                              rect=(left, bottom, right, top))
383  
384  ->     return kwargs
(Pdb) p kwargs
{'top': 0.43099934895833336, 'right': 0.97750000000000004, 'bottom': 0.56900065104166675, 'left': 0.16939941406250003}
```

I guess this in turns comes from what happens in `auto_adjust_subplotpars`, in the output of `_get_bottom` and `_get_top`:

```
(Pdb) p ax_bbox.ymin
0.3716666666666667
(Pdb) p tight_bbox.ymin
-0.16733398437499999
(Pdb) p ax_bbox.ymin - tight_bbox.ymin
0.53900065104166672
(Pdb) p tight_bbox.ymax
1.1673339843749999
(Pdb) p ax_bbox.ymax
0.6283333333333333
(Pdb) p tight_bbox.ymax - ax_bbox.ymax
0.53900065104166661
```

I don't know what `ax_bbox` and `tight_bbox` exactly are.

I just guess the algorithm to determine the figure properties do not play well with the data I want to plot. Can this be considered a bug ?

### PR DESCRIPTION
**Pull request by josecv**
For #5456 added a check to keep margins < .5 to prevent an Exception

### CODE REVIEW

File: lib/matplotlib/tests/test_tightlayout.py [hunk 1]
```diff
@@ -253,3 +253,17 @@ def _subplots():
                 child.set_visible(False)
 
     plt.tight_layout()
+
+
+def test_long_label():
+  '''
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_tightlayout.py]: Please use 4-space indent.

File: lib/matplotlib/tight_layout.py [hunk 2]
```diff
@@ -165,22 +165,28 @@ def auto_adjust_subplotpars(fig, renderer,
 
     # margins can be negative for axes with aspect applied. And we
     # append + [0] to make minimum margins 0
+    # We must also ensure that maximum margins are .5 so that we do not
+    # end up with bottom >= top or left >= right
 
     if not margin_left:
         margin_left = max([sum(s) for s in hspaces[::cols + 1]] + [0])
         margin_left += pad_inches / fig_width_inch
+        margin_left = 0.5 if margin_left > 0.5 else margin_left
```
Comment (mdboom) [inline lib/matplotlib/tight_layout.py]: It's more efficient (and IMHO more readable) to write this as:

```
margin_left = max(0.5, margin_left)
```

Comment (QuLogic) [inline lib/matplotlib/tight_layout.py]: I think that's `min`, not `max`...

**Comment by josecv** [comment]
Ok, thanks, does this look better?

File: lib/matplotlib/tight_layout.py (hunk 2 above)
Comment (mdboom) [inline lib/matplotlib/tight_layout.py]: Indeed.

**Comment by josecv** [comment]
Hurm the tests seem to be failing with a latex error on windows, could it be the ylabel is too long for latex' tastes?

**Comment by mdboom** [comment]
It seems like the test failure is unrelated.  Has anyone seen that before?  (Cc: @janschulz)

**Comment by janschulz** [comment]
It seems latex is run in that test (if that's not ok, then the rest of my explanation is wrong :-)), but I'm not sure if appveyor has latex installed (I'm pretty sure it has not). Usually there are test decorators which test for latex and skip the test if no latex is available on the system: https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/tests/test_backend_ps.py#L22-L25 and used then in https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/tests/test_backend_ps.py#L72-L78

```
====================================================================== 
ERROR: Test for 5456. When the label was too long we'd get a ValueError that 
---------------------------------------------------------------------- 
Traceback (most recent call last):
  File "C:\conda\envs\test-environment\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "c:\projects\matplotlib\lib\matplotlib\tests\test_tightlayout.py", line 266, in test_long_label
    plt.tight_layout()
  File "c:\projects\matplotlib\lib\matplotlib\pyplot.py", line 1282, in tight_layout
    fig.tight_layout(pad=pad, h_pad=h_pad, w_pad=w_pad, rect=rect)
  File "c:\projects\matplotlib\lib\matplotlib\figure.py", line 1857, in tight_layout
    rect=rect)
  File "c:\projects\matplotlib\lib\matplotlib\tight_layout.py", line 355, in get_tight_layout_figure
    pad=pad, h_pad=h_pad, w_pad=w_pad)
  File "c:\projects\matplotlib\lib\matplotlib\tight_layout.py", line 126, in auto_adjust_subplotpars
    tight_bbox_raw = union([ax.get_tightbbox(renderer) for ax in subplots])
  File "c:\projects\matplotlib\lib\matplotlib\axes\_base.py", line 3750, in get_tightbbox
    bb_xaxis = self.xaxis.get_tightbbox(renderer)
  File "c:\projects\matplotlib\lib\matplotlib\axis.py", line 1087, in get_tightbbox
    renderer)
  File "c:\projects\matplotlib\lib\matplotlib\axis.py", line 1070, in _get_tick_bboxes
    extent = tick.label1.get_window_extent(renderer)
  File "c:\projects\matplotlib\lib\matplotlib\text.py", line 961, in get_window_extent
    bbox, info, descent = self._get_layout(self._renderer)
  File "c:\projects\matplotlib\lib\matplotlib\text.py", line 352, in _get_layout
    ismath=False)
  File "c:\projects\matplotlib\lib\matplotlib\backends\backend_agg.py", line 230, in get_text_width_height_descent
    renderer=self)
  File "c:\projects\matplotlib\lib\matplotlib\texmanager.py", line 676, in get_text_width_height_descent
    dvifile = self.make_dvi(tex, fontsize)
  File "c:\projects\matplotlib\lib\matplotlib\texmanager.py", line 423, in make_dvi
    report))
RuntimeError: LaTeX was not able to process the following string:
'lp'
Here is the full report generated by LaTeX: 
 ```

**Comment by mdboom** [comment]
Ah -- thanks, @janschulz.  I don't LaTeX is *supposed* to be used in that test, but due to some sideeffect (probably of another test) it is.

@josecv: Can you try adding the `@cleanup` decorator to your new test?

**Comment by janschulz** [comment]
Cleanup is not enough, it has to skip if no Latex is installed.

**Comment by josecv** [comment]
Ah, perfect, thanks! Hopefully this should do the trick

**Comment by mdboom** [comment]
Ok -- I think the confusion is that the new test in this PR is not responsible for the failure, but it somehow triggering it in another test -- `test_backend_svg:test_determinism_tex`, which already has the `@needs_tex` decorator.

It isn't clear to me why the next test here needs to check for TeX -- it shouldn't be using it and isn't part of the traceback displayed on AppVeyor.

**Comment by janschulz** [comment]
I've seens ushc behaviour when I missed a `@cleanup` in the tests of ggplot, but this is here: https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/tests/test_backend_svg.py#L179-L184 -> I've no idea :-(

**Comment by josecv** [comment]
Huh Known Failure Did Not Fail:

```
====================================================================== 
ERROR: Test for 5456. No ValueError when using very long label. 
---------------------------------------------------------------------- 
Traceback (most recent call last):
  File "C:\conda\envs\test-environment\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "c:\projects\matplotlib\lib\matplotlib\testing\decorators.py", line 152, in wrapped_callable
    func(*args, **kwargs)
  File "c:\projects\matplotlib\lib\matplotlib\testing\decorators.py", line 67, in failer
    raise KnownFailureDidNotFailTest(msg)
KnownFailureDidNotFailTest: This test needs a Tex installation  
----------------------------------------------------------------------
```

Is it possible that the cleanup decorator on its own is allowing the test to pass?

**Comment by WeatherGod** [comment]
Well, more specifically, it did fail, but for the wrong reasons.

I am wondering if usetex=True is leaking from somewhere else?

**Comment by josecv** [comment]
Hmm, strange - hopefully then leaving the test with only `@cleanup` is going to do the trick

**Comment by josecv** [comment]
Oh my, bizarrely it's the Travis build that's failed now, though it seems to be an unrelated failure in a single test (namely `--test-args=pep8` with python 3.5):

```
The command "curl -s -o python-3.5.tar.bz2 https://s3.amazonaws.com/travis-python-archives/binaries/$(lsb_release -is | tr "A-Z" "a-z")/$(lsb_release -rs)/$(uname -m)/python-3.5.tar.bz2" failed and exited with 18 during .

Your build has been stopped.
```

**Comment by WeatherGod** [comment]
And now everything passes. Go figure.

Just one thought on this. This approach solves the problem for the fully automatic case, but what about the case where the user specified some of the margins?

**Comment by josecv** [comment]
That's a good point - the min statements could easily be pulled out from the `if not margin_*:` statements, would that be the best thing to do?
The user would find that their margins are being ignored, but at least there won't be an exception. Alternatively, tight_layout could itself produce an error when fed "bad" (out of bounds) margins.

**Comment by tacaswell** [comment]
https://github.com/matplotlib/matplotlib/pull/6167 is another PR addressing the same issue.  I left some comments on the other PR that I think apply here as well.

@kskod @josecv Are you both in CSC D01?  If so can you please collaborate on this?

**Comment by anntzer** [comment]
I am closing this as I believe we should instead have a more general solution which raises whenever the padding/spacing parameters make tight-layouting impossible (#8062, #8201, #8202) (or handles correctly such cases).
Feel free to ask for a reopen if you disagree.

**Pull request closed without merging**